typescript/xstate/src/lib.ts

```diff
@@ -0,0 +1,360 @@
+  }
+
+  return createXActor(logic, {id: "root", ...options, parent, snapshot} as any);
+}
+
+const actorMethods = <TLogic extends AnyStateMachine>(path: string, logic: TLogic) => {
+  const api = xStateApi(path)
+
+  return {
+    create: async (ctx: restate.RpcContext, systemName: string, request?: { input?: InputFrom<TLogic> }): Promise<Snapshot<unknown>> => {
+      ctx.clear("snapshot")
+      ctx.clear("events")
+      ctx.clear("children")
+
+      const root = (await createActor(ctx, api, systemName, logic, {
+        input: request?.input,
+      })).start();
+
+      ctx.set("snapshot", root.getPersistedSnapshot())
+
+      return root.getPersistedSnapshot()
+    },
+    send: async (ctx: restate.RpcContext, systemName: string, request?: { scheduledEvent?: SerialisableScheduledEvent, source?: SerialisableActorRef, target?: SerialisableActorRef, event: AnyEventObject }): Promise<Snapshot<unknown> | undefined> => {
+      if (!request) {
+        throw new TerminalError("Must provide a request")
+      }
+
+      if (request.scheduledEvent) {
+        const events = await ctx.get<{ [key: string]: SerialisableScheduledEvent }>("events") ?? {}
+        const scheduledEventId = createScheduledEventId(request.scheduledEvent.source, request.scheduledEvent.id)
+        if (!(scheduledEventId in events)) {
+          console.log("Received now cancelled event", scheduledEventId, "for target", request.target)
+          return
+        }
+        if (events[scheduledEventId].uuid !== request.scheduledEvent.uuid) {
+          console.log("Received now replaced event", scheduledEventId, "for target", request.target)
+          return
+        }
+        delete events[scheduledEventId]
+        ctx.set("events", events)
+      }
+
+      const root = (await createActor(ctx, api, systemName, logic)).start();
+
+
+      let actor;
+      if (request.target) {
+        actor = (root.system as RestateActorSystem<any>).actor(request.target.sessionId)
+        if (!actor) {
+          throw new TerminalError(`Actor ${request.target.id} not found; it may have since stopped`)
+        }
+      } else {
+        actor = root
+      }
+
+      (root.system as RestateActorSystem<any>)._relay(request.source, actor, request.event)
+
+      const nextSnapshot = root.getPersistedSnapshot()
+      ctx.set("snapshot", nextSnapshot)
+
+      return nextSnapshot
+    },
+    snapshot: async (ctx: restate.RpcContext, systemName: string): Promise<Snapshot<unknown>> => {
+      const root = (await createActor(ctx, api, systemName, logic));
+
+      return root.getPersistedSnapshot()
+    },
+  }
+}
+
+export const bindXStateRouter = <
+  TLogic extends AnyStateMachine,
+>(server: restate.RestateServer, path: string, logic: TLogic): restate.RestateServer => {
+  return server
+    .bindKeyedRouter(path, restate.keyedRouter(actorMethods(path, logic)))
+    .bindRouter(`${path}.promises`, restate.router(promiseMethods(path, logic)))
```

Comment:
This is a separate service to offload long-running promise execution to, so that the main event loop isnt blocked (restate processes events in a keyed service in series). this is more of a restate hack right now, we would like a way to have non locking methods on the same service